File: php/src/TruthMarkClient.php
<?php

namespace TruthMark\SDK;

class TruthMarkClient
{
    private string $baseUrl;
    private ?string $apiKey;
    private int $timeout;

    public function __construct(array $config = [])
    {
        $this->baseUrl = $config['base_url'] ?? 'http://localhost:8000';
        $this->apiKey = $config['api_key'] ?? null;
        $this->timeout = $config['timeout'] ?? 30;
    }

    /**
     * Embed an invisible watermark into an image
     *
     * @param string $imagePath Path to input image
     * @param string $message Text to embed (max 500 chars)
     * @return array Encode result with metadata
     * @throws \Exception
     */
    public function encode(string $imagePath, string $message): array
    {
        if (!file_exists($imagePath)) {
            throw new \Exception("Image file not found: {$imagePath}");
        }

        $ch = curl_init();
        
        $data = [
            'file' => new \CURLFile($imagePath),
            'message' => $message
        ];

        $headers = [];
        if ($this->apiKey) {
            $headers[] = "Authorization: Bearer {$this->apiKey}";
        }

        curl_setopt_array($ch, [
            CURLOPT_URL => "{$this->baseUrl}/v1/encode",
            CURLOPT_POST => true,
            CURLOPT_POSTFIELDS => $data,
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_HTTPHEADER => $headers,
            CURLOPT_TIMEOUT => $this->timeout,
        ]);

        $response = curl_exec($ch);
        $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
        
        if (curl_errno($ch)) {
            $error = curl_error($ch);
            curl_close($ch);
            throw new \Exception("cURL error: {$error}");
        }
        
        curl_close($ch);

        if ($httpCode !== 200) {
            throw new \Exception("API Error: HTTP {$httpCode}");
        }

        $result = json_decode($response, true);
        if (json_last_error() !== JSON_ERROR_NONE) {
            throw new \Exception("Failed to parse JSON response");
        }

        return $result;
    }

    /**
     * Extract watermark from an image
     *
     * @param string $imagePath Path to watermarked image
     * @return array Decode result with message and confidence
     * @throws \Exception
     */
    public function decode(string $imagePath): array
    {
        if (!file_exists($imagePath)) {
            throw new \Exception("Image file not found: {$imagePath}");
        }

        $ch = curl_init();
        
        $data = [
            'file' => new \CURLFile($imagePath)
        ];

        $headers = [];
        if ($this->apiKey) {
            $headers[] = "Authorization: Bearer {$this->apiKey}";
        }

        curl_setopt_array($ch, [
            CURLOPT_URL => "{$this->baseUrl}/v1/decode",
            CURLOPT_POST => true,
            CURLOPT_POSTFIELDS => $data,
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_HTTPHEADER => $headers,
            CURLOPT_TIMEOUT => $this->timeout,
        ]);

        $response = curl_exec($ch);
        $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
        
        if (curl_errno($ch)) {
            $error = curl_error($ch);
            curl_close($ch);
            throw new \Exception("cURL error: {$error}");
        }
        
        curl_close($ch);

        if ($httpCode !== 200) {
            throw new \Exception("API Error: HTTP {$httpCode}");
        }

        $result = json_decode($response, true);
        if (json_last_error() !== JSON_ERROR_NONE) {
            throw new \Exception("Failed to parse JSON response");
        }

        return $result;
    }
}

File: kotlin/TruthMarkClient.kt
package com.truthmark.sdk

import okhttp3.*
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.RequestBody.Companion.asRequestBody
import com.google.gson.Gson
import java.io.File
import java.io.IOException
import java.util.concurrent.TimeUnit

data class TruthMarkConfig(
    val baseUrl: String = "http://localhost:8000",
    val apiKey: String? = null,
    val timeoutSeconds: Long = 30
)

data class EncodeResult(
    val status: String,
    val metadata: Metadata,
    val download_url: String
) {
    data class Metadata(
        val psnr: Double,
        val bits_embedded: Int
    )
}

data class DecodeResult(
    val found: Boolean,
    val message: String?,
    val confidence: Double
)

class TruthMarkClient(config: TruthMarkConfig = TruthMarkConfig()) {
    private val baseUrl = config.baseUrl
    private val apiKey = config.apiKey
    private val httpClient: OkHttpClient
    private val gson = Gson()

    init {
        httpClient = OkHttpClient.Builder()
            .connectTimeout(config.timeoutSeconds, TimeUnit.SECONDS)
            .readTimeout(config.timeoutSeconds, TimeUnit.SECONDS)
            .build()
    }

    /**
     * Embed an invisible watermark into an image
     * @param imagePath Path to input image
     * @param message Text to embed (max 500 chars)
     * @return EncodeResult with metadata and download URL
     */
    @Throws(IOException::class)
    fun encode(imagePath: String, message: String): EncodeResult {
        val imageFile = File(imagePath)
        if (!imageFile.exists()) {
            throw IOException("Image file not found: $imagePath")
        }

        val requestBody = MultipartBody.Builder()
            .setType(MultipartBody.FORM)
            .addFormDataPart(
                "file",
                imageFile.name,
                imageFile.asRequestBody("image/*".toMediaTypeOrNull())
            )
            .addFormDataPart("message", message)
            .build()

        val requestBuilder = Request.Builder()
            .url("$baseUrl/v1/encode")
            .post(requestBody)

        apiKey?.let {
            requestBuilder.addHeader("Authorization", "Bearer $it")
        }

        val response = httpClient.newCall(requestBuilder.build()).execute()

        if (!response.isSuccessful) {
            throw IOException("API Error: ${response.code}")
        }

        val responseBody = response.body?.string()
            ?: throw IOException("Empty response body")

        return gson.fromJson(responseBody, EncodeResult::class.java)
    }

    /**
     * Extract watermark from an image
     * @param imagePath Path to watermarked image
     * @return DecodeResult with extracted message
     */
    @Throws(IOException::class)
    fun decode(imagePath: String): DecodeResult {
        val imageFile = File(imagePath)
        if (!imageFile.exists()) {
            throw IOException("Image file not found: $imagePath")
        }

        val requestBody = MultipartBody.Builder()
            .setType(MultipartBody.FORM)
            .addFormDataPart(
                "file",
                imageFile.name,
                imageFile.asRequestBody("image/*".toMediaTypeOrNull())
            )
            .build()

        val requestBuilder = Request.Builder()
            .url("$baseUrl/v1/decode")
            .post(requestBody)

        apiKey?.let {
            requestBuilder.addHeader("Authorization", "Bearer $it")
        }

        val response = httpClient.newCall(requestBuilder.build()).execute()

        if (!response.isSuccessful) {
            throw IOException("API Error: ${response.code}")
        }

        val responseBody = response.body?.string()
            ?: throw IOException("Empty response body")

        return gson.fromJson(responseBody, DecodeResult::class.java)
    }
}
